### ISSUE TITLE
Constructor call with () => void as second type param triggers func-call-spacing

### ISSUE THREAD

**Issue by greim** [typescript-eslint/typescript-eslint#364]
**What code were you trying to parse?**

```ts
const invalidators = new Map<any, () => void>();
```

**What did you expect to happen?**

Not run afoul of eslint's `func-call-spacing` rule.

**What actually happened?**

Runs afoul of eslint's `func-call-spacing` rule.

Lint message: "Unexpected newline between function name and paren. eslint(func-call-spacing)"

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `1.4.2` |
| `TypeScript`                | `3.3.3` |
| `ESLint`                    | `5.14.1` |
| `node`                      | `8.x` |
| `npm`                       | `8.x` |

Here's some variants:
```js
const invalidators = new Map<any, any>(); // ok
const invalidators = new Map<() => void, any>(); // ok
const invalidators = new Map<any, typeof thing>(); // ok
const invalidators = new Map<any, () => void>(); // triggers (but shouldn't)
const invalidators = new Map<any, (s: string) => string>(); // triggers (but shouldn't)
const invalidators = new Map <any, () => void>(); // triggers
const invalidators = new Map<any, () => void> (); // triggers
```

**Comment by greim** [comment]
The reason I created this issue in the "parser" category is I assume TypeScript erases the types and emits a newline in that spot, which is what's triggering `func-call-spacing`. Also, this still happens if I swap in the rule `no-space-func`. Thus I don't think this is a rule-level problem.

**Comment by bradzacher** [comment]
You would be incorrect in your assumption - the parser does not scrub types from the AST, as this would be counterproductive for linting typescript features.

The reason this occurs is because the base eslint rules do not have handling for typescript syntax.
Most of them "just work" because they are based off of the AST, or don't inspect the source code at places where typescript adds features.

However some of them (like both [`func-call-spacing`](https://github.com/eslint/eslint/blob/master/lib/rules/func-call-spacing.js) and [`no-space-func`](https://github.com/eslint/eslint/blob/master/lib/rules/no-spaced-func.js)) inspect the source code's tokens and assert that is has a specific structure.

For example, `func-call-spacing` (note that I'm doing this by hand, so some values are probably *slightly* incorrect, but the ideas are correct).
 
1) grab the `NewExpression` node (= `new Map<any, (s: string) => string>()`)
2) grab the last token of the expression (= `)`)
3) grab the last token of the callee (= `Map`)
4) grab the first opening parenthesis token between (2) and (3) (which will get the `(` within the generic parameter, ***NOT*** the second to last character in the expression)
5) get the token before (4) (= `any`)
6) get the source code between (4) and (5) (= `, `)
7) check (6) for a space character (which will find one, and thus report an error)

`no-spaced-func` works in a similar way, I believe.

Following this process through for the working examples, you can see why they don't trigger - specifically because step (4) will get the correct parenthesis, so the rest of the steps will work as intended.

We ***cannot*** patch the base eslint rule, because the base eslint rules are designed and implemented around the base Javascript spec, and adding TS exceptions could cause bugs in normal JS use cases.
So we would probably need to extend these base rules by creating them as rules in our plugin in order to make them work correctly, so that we can implement the typescript specific logic.

------

As a workaround for now:
you can either disable the rule with an eslint comment, or hoist the type outside of the generic.

**Comment by greim** [comment]
Thanks for this informative analysis!
